public/serviceWorker.js

```diff
@@ -0,0 +1,366 @@
+            return decodedAndParsedJWTBody.exp;
+        })();
+
+        /**
+         * If access token is expired, then we have to refresh it.
+         * If the user has a valid refresh token, then this will succeed.
+         * We will do this right now before handling the current request to make it as seamless as possible,
+         * the user won't even notice that it has been refreshed.
+         */
+        if (accessTokenExpiration < getTimestampInSeconds()) {
+            accessToken = await refreshTokens(event)
+        }
+
+        /**
+         * If Access Token a string and is not ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED, then we can assume that it's a real JWT.
+         * All we have to do now is to modify the original request and add our 'Grammatip-Authorization' header.
+         */
+        if (typeof accessToken === 'string' && accessToken !== ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED) {
+            const modifiedRequest = addAuthHeaderToRequest(event.request, accessToken);
+
+            /**
+             * If url is an "unauthorize"
+             */
+            if (isUnauthorizeUrls(url)) {
+                return HandleUnAuthorizes(event, modifiedRequest);
+            }
+
+            return fetch(modifiedRequest);
+        }
+    }
+
+    /**
+     * Return unmodified, backend will handle access control and redirection if necessary
+     */
+    return fetch(event.request);
+}
+
+/**
+ * Refresh token. Get a new accessToken.
+ * Call this if old token has become invalid.
+ *
+ * @param {Event} event
+ * @return {String} Either a valid AccessToken or ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED, depending on the validity of the refresh token.
+ */
+async function refreshTokens(event) {
+    /**
+     * Extra check before we start a refresh.
+     * Check if there already is one started.
+     * If there is then wait for it to complete.
+     */
+    if (refreshingTokensLock.isLocked) {
+        await refreshingTokensLock.waitForUnlock()
+
+        return await localforage.getItem(ACCESS_TOKEN)
+    }
+
+    /**
+     * A AbortController is nessesary to be able to abort the fetch request if it takes too long to complete.
+     *
+     * Without the AbortController there would be chance that we would
+     * get a response between AutoUnlock triggering in the ExecutionLock
+     * and the fetch request timeouting at 300 seconds.
+     * If this happens we would run into the Automatic reuse detection triggering in the backend.
+     */
+    const controller = new AbortController();
+
+    refreshingTokensLock.lock(async () => {
+        // Aborting the fetch request.
+        controller.abort()
+        await localforage.setItem(ACCESS_TOKEN, ACCESS_TOKEN_CONFIRMED_UNAUTHORIZED);
+    });
+
+    let accessToken;
+    let refreshResponse = await fetch('/api/auth/refresh', {
+        method: 'POST',
+        signal: controller.signal
+    });
+    event.waitUntil(refreshResponse);
```

Comment:
Won't this overwrite our current refresh token api code?

Also use PUBLIC_API_URL from .env

Reply:
No, this uses the refresh token api code. Acts as a middleware between client and server.  

Decided it was not worth it right now to add url.